mathics/core/expression.py

```diff
@@ -1165,6 +1250,132 @@ def sameQ(self, other: BaseElement) -> bool:
             for element, oelement in zip(self._elements, other.get_elements())
         )
 
+    def sequences(self):
+        cache = self._cache
+        if cache:
+            seq = cache.sequences
+            if seq is not None:
+                return seq
+
+        return self._rebuild_cache().sequences
+
+    def set_head(self, head):
+        self._head = head
+        self._cache = None
+
+    def set_element(self, index: int, value):
+        """
+        Update element[i] with value
+        """
+        elements = list(self._elements)
+        elements[index] = value
+        self._elements = tuple(elements)
+        self._cache = None
+
+    def set_reordered_elements(
```

Comment:
This is used only in sort.

Reply:
Here I have "inlined" this. But here's the rationale and this is kind of important to me. 

My first aim was to reduce the method exposure of an Expression object since right now this is so large that I find it is a 3 line and `sort` was about 5 lines. With such small custom stuff, adding a comment before the 3 lines seems sufficient for clarity. 

And for the fans of squeezing the last ounce of performance, it occured to me that since this code is a bit heavy weight we should check to see if there is something to sort. That is there is more than 1 element. Otherwise we do the extra work to convert a tuple into a list, set up a list sorting key, do the sort, convert the list back into a tuple, and then possibly do something with a cache. 

Testing shows indeed there are a number of functions with zero and 1 argument.

Reply:
@rocky @mmatera maybe the sorting could be made after the evaluation has finished?

Reply:
@mmatera @TiagoCavalcante 

The focus here was on _clarity_ not figuring out how to improve it. 

What got us here was trying to describe in a high-level what is involved when we evaluate. And there was a bit of a mess that I couldn't do the first proposal without making things possibly more complicated.

Discussing this proposal in that big picture discussion is to my mind preferable to looking at a PR and discussing that here which is tangential.

So @mmatera I'd suggest looking at the overall steps and in that describe how you think things should change. Then if you think this still works, let's discuss.